File: client/python/lila_client/world_model.py
"""līlā Python Client — World Model (local scene graph).

Mirrors the browser client's WorldModel for entity tracking and spatial queries.
Used by both the ImGui renderer and the local agency system.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class WorldEntity:
    """Client-side entity record with server reference and local position."""

    id: str
    type: str = ""          # ANIMAL, PLANT, TREE, INSECT, BIRD, MICROORGANISM
    species: str = ""

    # Server reference position (gravity well for reconciliation)
    ref_x: float = 0.0
    ref_z: float = 0.0

    # Client-agency position (where the entity actually is in local sim)
    x: float = 0.0
    z: float = 0.0

    # Discrete state from server
    state: str = "IDLE"

    # Drive values from server intent packet
    drive: dict[str, float] = field(default_factory=dict)

    # Motion latent vector (4D, from motor model)
    motion_latent: list[float] = field(default_factory=lambda: [0.0, 0.0, 0.0, 0.0])

    # Eligibility flags from server
    can_consume: bool = False
    can_predate: bool = False
    can_pollinate: bool = False
    repro_eligible: bool = False
    can_drink: bool = False
    spread_eligible: bool = False

    # Skeleton ID (for future 3D client)
    skeleton_id: Optional[str] = None

    # Local agency state
    target_x: float = 0.0
    target_z: float = 0.0
    has_target: bool = False
    velocity_x: float = 0.0
    velocity_z: float = 0.0
    speed: float = 1.0

    # Facing angle (radians, smoothed via lerp toward velocity direction)
    facing_angle: float = 0.0

    # Acknowledgment tracking
    ack_received: bool = False

    # Reconciliation queue — target positions enqueued by reconcile(),
    # consumed smoothly by the agency system at 60fps.
    _reconcile_queue: list[tuple[float, float]] = field(default_factory=list)
    _reconcile_idx: int = 0  # index of current target in the queue

    # ── Per-entity sync personality (deterministic from ID) ──
    # Spread out reconciliation so entities don't all nudge at once.
    _sync_phase: int = 0          # 0..3 — which frame within sync cycle to react
    _sync_speed: float = 0.7      # 0.4..1.0 multiplier on nudge intensity
    _last_reconciled_tick: int = -10  # force reconcile on first ticks

    @property
    def is_alive(self) -> bool:
        return self.state not in ("DEAD", "DYING", "DORMANT")

    @property
    def is_mobile_consumer(self) -> bool:
        return self.type in ("ANIMAL", "BIRD", "INSECT")

    def distance_to(self, other: WorldEntity) -> float:
        dx = self.x - other.x
        dz = self.z - other.z
        return math.sqrt(dx * dx + dz * dz)

    def dist_sq_to(self, other: WorldEntity) -> float:
        dx = self.x - other.x
        dz = self.z - other.z
        return dx * dx + dz * dz


class WorldModel:
    """Client-side world model — local scene graph with spatial queries."""

    def __init__(self):
        self.entities: dict[str, WorldEntity] = {}
        self.species_defs: dict[str, dict] = {}
        self.water_sources: list[dict] = []
        # Moisture grid for heatmap rendering (GRID_SIZE × GRID_SIZE)
        from .constants import GRID_SIZE
        self.moisture = [0.65] * (GRID_SIZE * GRID_SIZE)
        self._grid_size = GRID_SIZE

    # ─── Entity Management ──────────────────────────────────────────────

    def apply_update(self, u: dict) -> WorldEntity:
        """Add or update an entity from a server tick packet."""
        eid = u["id"]
        ent = self.entities.get(eid)

        if not ent:
            info = _infer_entity_type_from_id(eid)
            ent = WorldEntity(
                id=eid, type=info["type"], species=info["species"],
            )
            ref_pos = u.get("ref_position", [0, 0, 0])
            ent.ref_x = ref_pos[0]
            ent.ref_z = ref_pos[2]
            ent.x = ref_pos[0]
            ent.z = ref_pos[2]
            _init_sync_personality(ent)
            self.entities[eid] = ent

        # Update server reference position (gravity well)
        if "ref_position" in u:
            pos = u["ref_position"]
            ent.ref_x = pos[0]
            ent.ref_z = pos[2]

        # State and drives
        if "state" in u:
            ent.state = u["state"]
        if "drive" in u:
            ent.drive.update(u["drive"])
        if "motion_latent" in u:
            ent.motion_latent = list(u["motion_latent"])

        # Eligibility flags
        ent.can_consume = bool(u.get("_can_consume", False))
        ent.can_predate = bool(u.get("_can_predate", False))
        ent.can_pollinate = bool(u.get("_can_pollinate", False))
        ent.repro_eligible = bool(u.get("_repro_eligible", False))
        ent.can_drink = bool(u.get("_can_drink", False))
        ent.spread_eligible = bool(u.get("_spread_eligible", False))

        # Acknowledgment tracking
        ent.ack_received = bool(u.get("_ack", False))

        return ent

    def apply_spawn(self, s: dict) -> WorldEntity:
        """Spawn a new entity from server spawn packet."""
        ref_pos = s["ref_position"]
        ent = WorldEntity(
            id=s["id"],
            type=s.get("type", ""),
            species=s.get("species", ""),
            ref_x=ref_pos[0], ref_z=ref_pos[2],
            x=ref_pos[0], z=ref_pos[2],
        )
        if "state" in s:
            ent.state = s["state"]
        if "drive" in s:
            ent.drive.update(s["drive"])
        if "skeleton_id" in s:
            ent.skeleton_id = s["skeleton_id"]
        _init_sync_personality(ent)
        self.entities[ent.id] = ent
        return ent

    def apply_removal(self, eid: str) -> Optional[WorldEntity]:
        """Remove an entity. Returns the removed entity or None."""
        return self.entities.pop(eid, None)

    # ─── Spatial Queries (for agency logic) ──────────────────────────────

    def find_nearest(
        self, x: float, z: float, types: list[str], exclude_id: str | None = None,
    ) -> Optional[WorldEntity]:
        """Find nearest alive entity of given type(s)."""
        best_dist_sq = float("inf")
        best_ent = None
        for ent in self.entities.values():
            if not ent.is_alive:
                continue
            if exclude_id and ent.id == exclude_id:
                continue
            if ent.type not in types:
                continue
            d2 = (x - ent.x) ** 2 + (z - ent.z) ** 2
            if d2 < best_dist_sq:
                best_dist_sq = d2
                best_ent = ent
        return best_ent

    def find_nearest_species(
        self, x: float, z: float, species_list: list[str], exclude_id: str | None = None,
    ) -> Optional[WorldEntity]:
        """Find nearest alive entity of given species."""
        best_dist_sq = float("inf")
        best_ent: WorldEntity | None = None
        for ent in self.entities.values():
            if not ent.is_alive:
                continue
            if exclude_id and ent.id == exclude_id:
                continue
            if ent.species not in species_list:
                continue
            d2 = (x - ent.x) ** 2 + (z - ent.z) ** 2
            if d2 < best_dist_sq:
                best_dist_sq = d2
                best_ent = ent
        return best_ent

    def find_nearest_mate(self, ent: WorldEntity) -> Optional[WorldEntity]:
        """Find nearest alive entity of the same species."""
        best_dist_sq = float("inf")
        best_ent: WorldEntity | None = None
        for other in self.entities.values():
            if not other.is_alive:
                continue
            if other.id == ent.id:
                continue
            if other.species != ent.species:
                continue
            d2 = ent.dist_sq_to(other)
            if d2 < best_dist_sq:
                best_dist_sq = d2
                best_ent = other
        return best_ent

    def get_species_def(self, species_id: str) -> dict:
        """Get species definition by ID."""
        return self.species_defs.get(species_id, {})

    def find_nearest_water(self, x: float, z: float) -> Optional[dict]:
        """Find nearest non-dry water source."""
        best_dist_sq = float("inf")
        best_source = None
        for ws in self.water_sources:
            if (ws.get("water_level", 1.0) or 1.0) < 0.05:
                continue
            d2 = (x - ws["position"][0]) ** 2 + (z - ws["position"][2]) ** 2
            if d2 < best_dist_sq:
                best_dist_sq = d2
                best_source = ws
        return best_source

    def get_alive_of_type(self, entity_type: str) -> list[WorldEntity]:
        """Get all alive entities of a given type."""
        return [e for e in self.entities.values() if e.is_alive and e.type == entity_type]

    # ─── Voxel / Environment ──────────────────────────────────────────────

    def apply_voxel_deltas(self, deltas: dict) -> None:
        """Apply moisture voxel deltas from server."""
        if not deltas or "moisture" not in deltas:
            return
        for coord_str, val in deltas["moisture"].items():
            parts = list(map(int, coord_str.split(",")))
            gx, gz = parts[0], parts[2]
            if 0 <= gx < self._grid_size and 0 <= gz < self._grid_size:
                self.moisture[gz * self._grid_size + gx] = val

    def apply_water_sources(self, sources: list[dict]) -> None:
        """Update water source positions."""
        self.water_sources = sources or []


def _init_sync_personality(ent: WorldEntity) -> None:
    """Initialize per-entity sync personality (deterministic from ID).

    Spreads out reconciliation so entities don't all nudge at once.
    _sync_phase: 0..3 — which frame within a sync cycle this entity reacts
    _sync_speed: 0.4..1.0 — multiplier on nudge intensity
    """
    seed = _hash_id(ent.id)
    ent._sync_phase = seed % 4               # 0, 1, 2, or 3
    ent._sync_speed = 0.4 + (seed % 60) / 100  # 0.40 .. 0.99
    ent._last_reconciled_tick = -10


def _hash_id(eid: str) -> int:
    """Deterministic hash from entity ID string → positive integer."""
    h = 0
    for c in eid:
        h = ((h << 5) - h + ord(c)) & 0xFFFFFFFF
    return abs(h)


def _infer_entity_type_from_id(eid: str) -> dict[str, str]:
    """Infer entity type from ID prefix (fallback when server data is sparse)."""
    if eid.startswith("deer"):
        return {"type": "ANIMAL", "species": "deer"}
    if eid.startswith("wolf"):
        return {"type": "ANIMAL", "species": "wolf"}
    if eid.startswith("bird") or eid.startswith("songbird"):
        return {"type": "BIRD", "species": "songbird"}
    if eid.startswith("butterfly"):
        return {"type": "INSECT", "species": "monarch"}
    if eid.startswith("oak"):
        return {"type": "TREE", "species": "meadow_oak"}
    if eid.startswith("grass"):
        return {"type": "PLANT", "species": "meadow_grass"}
    if eid.startswith("flower"):
        return {"type": "PLANT", "species": "wildflower"}
    if eid.startswith("mushroom") or eid.startswith("fungus"):
        return {"type": "MICROORGANISM", "species": "mushroom"}
    return {"type": "ANIMAL", "species": "unknown"}

File: client/browser/js/world-model.js
// ═══════════════════════════════════════════════════════
// līlā — World Model (Client-Side Scene Graph)
//
// Maintains the client's local view of all entities, species
// definitions from server, and provides query methods for
// agency logic (nearest food, nearest water, etc.).
// ═══════════════════════════════════════════════════════

import { GRID_SIZE } from './constants.js';

/**
 * Entity record in the client's local world model.
 * Tracks both server reference position and client-agency position.
 */
export class WorldEntity {
  constructor(id, type = '', species = '') {
    this.id = id;
    this.type = type;       // ANIMAL, PLANT, TREE, INSECT, BIRD, MICROORGANISM
    this.species = species;

    // Server reference position (gravity well for reconciliation)
    this.refX = 0;
    this.refZ = 0;

    // Client-agency position (where the entity actually is in local sim)
    this.x = 0;
    this.z = 0;

    // Discrete state from server
    this.state = 'IDLE';

    // Drive values from server intent packet
    this.drive = {};

    // Motion latent vector (4D, from motor model)
    this.motionLatent = [0, 0, 0, 0];

    // Eligibility flags from server
    this.canConsume = false;
    this.canPredatate = false;
    this.canPollinate = false;
    this.reproEligible = false;
    this.canDrink = false;
    this.spreadEligible = false;

    // Skeleton ID (for future Godot client)
    this.skeletonId = null;

    // Local agency state
    this.targetX = 0;
    this.targetZ = 0;
    this.hasTarget = false;
    this.velocityX = 0;
    this.velocityZ = 0;
    this.speed = 1.0;       // derived from species or default

    // Facing angle (radians, smoothed via lerp toward velocity direction)
    this.facingAngle = 0;

    // Acknowledgment tracking
    this.ackReceived = false;

    // Reconciliation queue — target positions enqueued by reconcile(),
    // consumed smoothly by the agency system at 60fps.
    this._reconcileQueue = [];
    this._reconcileIdx = 0;

    // ── Per-entity sync personality (deterministic from ID) ──
    // Spread out reconciliation so entities don't all nudge at once.
    // _syncPhase: 0..3 — which frame within a sync cycle this entity reacts
    // _syncSpeed: 0.4..1.0 multiplier on nudge intensity (slow / fast reactors)
    const seed = hashId(id);
    this._syncPhase = seed % 4;                  // 0, 1, 2, or 3
    this._syncSpeed = 0.4 + (seed % 60) / 100;   // 0.40 .. 0.99
    this._lastReconciledTick = -10;              // force reconcile on first ticks
  }

  /** Distance to another entity (2D xz-plane). */
  distanceTo(other) {
    const dx = this.x - other.x;
    const dz = this.z - other.z;
    return Math.sqrt(dx * dx + dz * dz);
  }

  /** Distance squared (for comparisons without sqrt). */
  distSqTo(other) {
    const dx = this.x - other.x;
    const dz = this.z - other.z;
    return dx * dx + dz * dz;
  }

  /** Is this entity alive and active? */
  get isAlive() {
    return !['DEAD', 'DYING', 'DORMANT'].includes(this.state);
  }

  /** Is this a mobile consumer (animal, bird, insect)? */
  get isMobileConsumer() {
    return ['ANIMAL', 'BIRD', 'INSECT'].includes(this.type);
  }
}

/**
 * World model — the client's local scene graph.
 * Provides entity lookup, spatial queries, and species definitions.
 */
export class WorldModel {
  constructor() {
    this.entities = new Map();       // id → WorldEntity
    this.speciesDefs = {};           // species_id → server-provided definition
    this.waterSources = [];          // [{ position, radius, water_level }]
    this.moisture = new Float32Array(GRID_SIZE * GRID_SIZE).fill(0.65);
  }

  // ─── Entity Management ──────────────────────────────

  /** Add or update an entity from a server tick packet. */
  applyUpdate(u) {
    let ent = this.entities.get(u.id);
    if (!ent) {
      const info = inferEntityTypeFromId(u.id);
      ent = new WorldEntity(u.id, info.type, info.species);
      // Initialize client position from server reference on first sight
      if (u.ref_position) {
        ent.x = u.ref_position[0];
        ent.z = u.ref_position[2];
        ent.refX = u.ref_position[0];
        ent.refZ = u.ref_position[2];
      }
      this.entities.set(u.id, ent);
    }

    // Update server reference position (gravity well)
    if (u.ref_position) {
      ent.refX = u.ref_position[0];
      ent.refZ = u.ref_position[2];
    }

    // Update state and drives from intent packet
    if (u.state) ent.state = u.state;
    if (u.drive) Object.assign(ent.drive, u.drive);
    if (u.motion_latent) ent.motionLatent = u.motion_latent;

    // Eligibility flags
    ent.canConsume = !!u._can_consume;
    ent.canPredatate = !!u._can_predate;
    ent.canPollinate = !!u._can_pollinate;
    ent.reproEligible = !!u._repro_eligible;
    ent.canDrink = !!u._can_drink;
    ent.spreadEligible = !!u._spread_eligible;

    // Acknowledgment tracking
    if (u._ack) {
      ent.ackReceived = true;
    } else {
      ent.ackReceived = false;
    }

    return ent;
  }

  /** Spawn a new entity from server spawn packet. */
  applySpawn(s) {
    const ent = new WorldEntity(
      s.id, s.type || '', s.species || ''
    );
    ent.refX = s.ref_position[0];
    ent.refZ = s.ref_position[2];
    // Client starts at server reference position for new entities
    ent.x = s.ref_position[0];
    ent.z = s.ref_position[2];
    if (s.state) ent.state = s.state;
    if (s.drive) Object.assign(ent.drive, s.drive);
    if (s.skeleton_id) ent.skeletonId = s.skeleton_id;
    this.entities.set(s.id, ent);
    return ent;
  }

  /** Remove an entity. */
  applyRemoval(id) {
    this.entities.delete(id);
  }

  // ─── Spatial Queries (for agency logic) ─────────────

  /** Find nearest alive entity of given type(s) from a position. */
  findNearest(x, z, types, excludeId = null) {
    let bestDist = Infinity;
    let bestEnt = null;
    for (const ent of this.entities.values()) {
      if (!ent.isAlive) continue;
      if (excludeId && ent.id === excludeId) continue;
      if (!types.includes(ent.type)) continue;
      const d2 = (x - ent.x) ** 2 + (z - ent.z) ** 2;
      if (d2 < bestDist) {
        bestDist = d2;
        bestEnt = ent;
      }
    }
    return bestEnt;
  }

  /** Find nearest alive entity of given species from a position. */
  findNearestSpecies(x, z, speciesList, excludeId = null) {
    let bestDist = Infinity;
    let bestEnt = null;
    for (const ent of this.entities.values()) {
      if (!ent.isAlive) continue;
      if (excludeId && ent.id === excludeId) continue;
      if (!speciesList.includes(ent.species)) continue;
      const d2 = (x - ent.x) ** 2 + (z - ent.z) ** 2;
      if (d2 < bestDist) {
        bestDist = d2;
        bestEnt = ent;
      }
    }
    return bestEnt;
  }

  /** Find nearest non-dry water source from a position. */
  findNearestWater(x, z) {
    let bestDist = Infinity;
    let bestSource = null;
    for (const ws of this.waterSources) {
      if ((ws.water_level ?? 1.0) < 0.05) continue; // dry
      const d2 = (x - ws.position[0]) ** 2 + (z - ws.position[2]) ** 2;
      if (d2 < bestDist) {
        bestDist = d2;
        bestSource = ws;
      }
    }
    return bestSource;
  }

  /** Find nearest mate (same species, different sex — approximated by ID). */
  findNearestMate(ent) {
    let bestDist = Infinity;
    let bestEnt = null;
    for (const other of this.entities.values()) {
      if (!other.isAlive) continue;
      if (other.id === ent.id) continue;
      if (other.species !== ent.species) continue;
      // Simple heuristic: different IDs of same species are potential mates
      const d2 = ent.distSqTo(other);
      if (d2 < bestDist) {
        bestDist = d2;
        bestEnt = other;
      }
    }
    return bestEnt;
  }

  /** Get all alive entities of a given type. */
  getAliveOfType(type) {
    const result = [];
    for (const ent of this.entities.values()) {
      if (ent.isAlive && ent.type === type) result.push(ent);
    }
    return result;
  }

  /** Get species definition from server-provided reference. */
  getSpeciesDef(speciesId) {
    return this.speciesDefs[speciesId] || null;
  }

  // ─── Voxel / Environment ────────────────────────────

  applyVoxelDeltas(deltas) {
    if (deltas && deltas.moisture) {
      for (const [coord, val] of Object.entries(deltas.moisture)) {
        const parts = coord.split(',').map(Number);
        const x = parts[0], z = parts[2];
        if (x >= 0 && x < GRID_SIZE && z >= 0 && z < GRID_SIZE) {
          this.moisture[z * GRID_SIZE + x] = val;
        }
      }
    }
  }

  applyWaterSources(sources) {
    this.waterSources = sources || [];
  }
}

// ─── Helpers ──────────────────────────────────────────

/** Deterministic hash from entity ID string → positive integer. */
function hashId(id) {
  let h = 0;
  for (let i = 0; i < id.length; i++) {
    h = ((h << 5) - h + id.charCodeAt(i)) | 0;
  }
  return Math.abs(h);
}

/** Infer entity type from ID prefix (fallback when server data is sparse). */
function inferEntityTypeFromId(id) {
  if (id.startsWith('deer'))      return { type: 'ANIMAL', species: 'deer' };
  if (id.startsWith('wolf'))      return { type: 'ANIMAL', species: 'wolf' };
  if (id.startsWith('bird') || id.startsWith('songbird')) return { type: 'BIRD', species: 'songbird' };
  if (id.startsWith('butterfly')) return { type: 'INSECT', species: 'monarch' };
  if (id.startsWith('oak'))       return { type: 'TREE',  species: 'meadow_oak' };
  if (id.startsWith('grass'))     return { type: 'PLANT', species: 'meadow_grass' };
  if (id.startsWith('flower'))    return { type: 'PLANT', species: 'wildflower' };
  if (id.startsWith('mushroom') || id.startsWith('fungus')) return { type: 'MICROORGANISM', species: 'mushroom' };
  return { type: 'ANIMAL', species: 'unknown' };
}
